Fix an HDFS data locality bug in case cluster node names are short host names

kimoonkim opened the pull request:
cc @ash211 @tnachen 

## What changes were proposed in this pull request?

Closes #290
 
## How was this patch tested?

Added unit tests.

Tested on a GKE cluster. The driver log now says tasks are dispatched `NODE_LOCAL` executors. Previously, it was `ANY` executors.

>
2017-05-22 20:07:28 INFO  KubernetesTaskSetManager:54 - Starting task 65.0 in stage 1.0 (TID 158, 10.0.0.11, executor 1, partition 65, NODE_LOCAL, 6089 bytes)
2017-05-22 20:07:28 INFO  KubernetesTaskSetManager:54 - Finished task 63.0 in stage 1.0 (TID 156) in 994 ms on 10.0.0.11 (executor 1) (63/93)
2017-05-22 20:07:28 INFO  KubernetesTaskSetManager:54 - Starting task 66.0 in stage 1.0 (TID 159, 10.0.0.12, executor 3, partition 66, NODE_LOCAL, 6089 bytes)
2017-05-22 20:07:28 INFO  KubernetesTaskSetManager:54 - Finished task 64.0 in stage 1.0 (TID 157) in 837 ms on 10.0.0.12 (executor 3) (64/93)

kimoonkim committed:
Fix an HDFS data locality bug in case cluster node names are not full host names

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManager.scala:
```diff
@@ -52,12 +56,28 @@ private[spark] class KubernetesTaskSetManager(
+private[kubernetes] class InetAddressUtil {
+
+  def getFullHostName(ipAddress: String): String = {
+    InetAddress.getByName(ipAddress).getCanonicalHostName
```
does this do a network call?  how long does the call take?  I'm wondering if it makes sense to cache these for a period of time

kimoonkim replied:
Yes, it does issue a network call to DNS. Caching is done by the `InetAddress` class. From the [java doc](https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html):

>InetAddress Caching
>
> The InetAddress class has a cache to store successful as well as unsuccessful host name resolutions.
By default, when a security manager is installed, in order to protect against DNS spoofing attacks, the result of positive host name resolutions are cached forever. When a security manager is not installed, the default behavior is to cache entries for a finite (implementation dependent) period of time. The result of unsuccessful host name resolution is cached for a very short period of time (10 seconds) to improve performance.

ash211 replied:
Ah nice, already does what I thought we might need!

ash211 reviewed (commented):
LGTM, though need to get tests passing before merge

kimoonkim committed:
Add a NOTE about InetAddress caching

ash211 committed:
Merge branch 'branch-2.1-kubernetes' into fix-hdfs-locality-bug-short-node-name

kimoonkim commented:
rerun unit tests please

ash211 reviewed (approved):
LGTM -- any thoughts @foxish / @tnachen / @mccheah ?

ash211 committed:
Merge branch 'branch-2.1-kubernetes' into fix-hdfs-locality-bug-short-node-name

ash211 commented:
Looking to merge this tomorrow if I don't hear otherwise

ash211 committed:
Merge branch 'branch-2.1-kubernetes' into fix-hdfs-locality-bug-short-node-name

ash211 committed:
Merge branch 'branch-2.1-kubernetes' into fix-hdfs-locality-bug-short-node-name

ash211 commented:
I think this is ready for merge -- will merge EOD unless someone gives a +1 earlier

tnachen commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManager.scala:
```diff
@@ -52,12 +56,30 @@ private[spark] class KubernetesTaskSetManager(
               s"$executorIP using cluster node IP $clusterNodeIP")
+            pendingTasksClusterNodeIP
+          } else {
+            val clusterNodeFullName = inetAddressUtil.getFullHostName(clusterNodeIP)
```
Nit: I wonder if we can reduce the nesting by returning the pendingTasks ArrayBuffer whenever it's non empty, so we can eliminate all the else blocks in this call.

kimoonkim replied:
Yes, that is possible. In fact, the first commit of the original PR was written in that style. But, there was a [review comment](https://github.com/apache-spark-on-k8s/spark/pull/216/files/b3855d651a7f639a25546fd0ca526b0bf5dbdbf7#r113525080) suggesting that the current form is the preferred scala style. PTAL the prior review comment and let me know what you think.

tnachen replied:
Ah ok, nvm then :)

tnachen commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesTaskSetManagerSuite.scala:
```diff
@@ -0,0 +1,117 @@
+  sched.backend = backend
+
+  test("Find pending tasks for executors using executor pod IP addresses") {
+    val taskSet = FakeTask.createTaskSet(3,
```
Move this to the next line

tnachen commented:
Just a nit for style, but otherwise LGTM

ash211 committed:
Merge branch 'branch-2.1-kubernetes' into fix-hdfs-locality-bug-short-node-name

kimoonkim commented:
Thanks @ash211 @tnachen for the reviews!

tangzhankun commented:
Hi,

There's issues of `getCanonicalHostName` used here in my k8s(v1.4.0) + calico cluster in physical machine:
1. This method costs _`5s`_ in the first run. Although there's cache after that, but the _`5s*numOfHosts`_ still matters when we run short-lived job.
2. The output of this method cannot guarantee full host name.

I ran below code in container using docker manually and "kubectl run" respectively, the output is different.
```java
import java.net.InetAddress;
public class test_inet {
  public static void main(String[] args) {
    try{
      System.out.println(InetAddress.getByName("172.16.3.70").getCanonicalHostName());
    }catch(Exception e) {
      e.printStackTrace();
    }
  }
}
```
When in k8s, it got host's ip address:
```bash
[root@r03s09 jobStartTimeBenchmark]# kubectl run -i -t test1 --image=172.16.3.78:4000/spark-driver:debug --restart="Never" --command -- /bin/bash
Waiting for pod default/test1 to be running, status is Pending, pod ready: false
Waiting for pod default/test1 to be running, status is Pending, pod ready: false
If you don't see a command prompt, try pressing enter.
bash-4.3# time java test_inet
172.16.3.70

real	0m5.199s
user	0m0.166s
sys	0m0.061s
```
But when manually run it, it reports host name.
```bash
[root@r03s09 jobStartTimeBenchmark]# docker run --rm -it 172.16.3.78:4000/spark-base:debug /bin/bash
bash-4.3# time java test_inet
r03s10

real	0m5.163s
user	0m0.140s
sys	0m0.070s
```
My questions are:
1. Do we must get full host name to search for pending task? When will the pending task hashmap have that in it?
2. Can k8s set full host name in pod spec for us to avoid the operations here?
@kimoonkim @tnachen

kimoonkim commented:
@tangzhankun Thanks for the report. And I am sorry that this bug impacted your cluster. Please see my answers below to your questions.
> Do we must get full host name to search for pending task? When will the pending task hashmap have that in it?

No, we don't need to get full host names to look up pending tasks. Not in most cases. For instance, the pending tasks hashmap will not be keyed by hosts at all in most cases. The stage input data will be associated with cluster nodes only when HDFS is used, which is not a common case for Spark on K8s.

It was a mistake in my code that we issued DNS requests even when there was no need. The DNS request code path was used as the last fallback when lookup with executor pod name or short host name of the cluster node do not return anything. Unfortunately, such lookup miss will happen a lot of times. So we end up issuing unnecessary DNS requests quite commonly. 

We fixed the issue in this PR by disabling the DNS request code path by default. We have a new release for both Spark 2.1 and 2.2 that has this PR's fix. Please get the new release if you haven't done it yet. And let us know if you still have a problem.

> Can k8s set full host name in pod spec for us to avoid the operations here?

A good question. It is possible to set full hostnames as cluster node names. Our internal cluster works this way. I also saw EC2 K8s clusters provisioned by `kops` also use full hostnames.

 So it is possible for cluster admin to avoid using short names. AFAIK, the short names were used only by GKE K8s clusters. For GKE, I couldn't find a way to force it to use full host names. So we tried to fix using the lookup code. And that's when we introduced the bug.

Hope I answered your questions. Please feel free to follow up with more questions or comments. Thanks.

tangzhankun commented:
@kimoonkim Thanks for the elaboration. Now I see the flag introduced in this https://github.com/apache-spark-on-k8s/spark/pull/412. This relieve the issue. But still two things unclear in my mind:

1. I'm using physical nodes only with short host name configured, this piece of `getFullHostName` seems doesn't work. Per my test program pasted above, It will return IP address in a container launched by k8s. While returns short host name when manually "docker run" it. 
**Why this piece of code have different output with the same input(physical node's IP address)? Was it caused by Calico DNS server or something else?**

2. If administrator configures both short host name `node1` and full name `node1.domain` on the nodes. For the case of application with HDFS input, 
**What will the pending tasks' hashmap be keyed with, "node1" or "node1.domain" or both or randomly? I'm trying to get the precise scenario to enable the flag `KUBERNETES_DRIVER_CLUSTER_NODENAME_DNS_LOOKUP_ENABLED`**

tangzhankun commented:
@kimoonkim @tnachen Any comments on my above questions?

tnachen commented:
For 1. you won't be getting the same results if you're manually running it in docker, since K8s does a lot more setup for you (replacing DNS resolv, mounting in different configs, etc). So the best test is to kubectl exec into a container run it manually there.

kimoonkim commented:
For 1. As @tnachen alluded, a vanilla docker and a K8s pod act differently in terms of network setup. K8s pods talk to kube-dns to resolve K8s service names, etc. For non-K8s entries, kube-dns can be set up to fall back to default DNS servers in charge of bare metal hosts. I wonder if your kube-dns is not configured for the fallback path. (I don't know much details about kube-dns :-))

Now let's talk about (2). The tasks hashmap will be keyed by "node1.domain" only if your setup is correct. Hadoop namenode/datanode always use the full hostnames. More, Hadoop requires your canonical hostname matches the full hostname. Since our [k8s-hdfs](https://github.com/apache-spark-on-k8s/kubernetes-HDFS) uses `hostNetwork`, this means the K8s cluster nodes should meet this requirement.

One easy way to see if your setup meets Hadoop requirement is to check if the output of `$ hostname` matches that of `$ hostname -f`. e.g.
```
$ hostname
node1.domain1
$  hostname -f
node1.domain1
```

The content of `/etc/hosts` file should look like the following. Below, it is very important for the full hostname appears first following the IP address. i.e. The short hostname should *not* come earlier than the full hostname in that line. It is also important that the first localhost entry does not have the hostname, short or long.
```
127.0.0.1       localhost.localdomain localhost
::1     localhost6.localdomain6 localhost6
192.168.6.157    node1.domain1 node1
```

With this, you might be able to use the flag `KUBERNETES_DRIVER_CLUSTER_NODENAME_DNS_LOOKUP_ENABLED`. But we agreed that it is better to avoid using the flag at all by setting up your K8s cluster nodes such that cluster node names match full hostnames. If your cluster admin can somehow make it happen, I think it'd be the best for your case. Maybe the above setup is all you need for that as well. Otherwise, you may want to look into how your `kubelet`s are launched.

tangzhankun commented:
@kimoonkim @tnachen Thanks for the explanation.
For question 1, Yes, my problem here might be caused by kube-dns. Need further debugging.
For question 2, I see when to use the flag. Thanks.

But back to the getFullHostName issue, the root cause is still not clear. Here's my environment (configure hosts with short host name is the common cases, right?)

```bash
[root@r03s09 ~]# ps aux|grep kubelet
root       6362  0.0  0.0 112648   968 pts/0    S+   13:24   0:00 grep --color=auto kubelet
root       9039  6.5  0.0 3715908 97592 ?       Ssl  Jul25 2249:04 /usr/bin/kubelet --kubeconfig=/etc/kubernetes/kubelet.conf --require-kubeconfig=true --pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true --network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin --cluster-dns=100.64.0.10 --cluster-domain=cluster.local --v=4
root     146612  1.2  0.0 179944 112308 ?       Ssl  Jul25 427:51 /usr/local/bin/kube-controller-manager --v=4 --address=0.0.0.0 --leader-elect --master=0.0.0.0:8080 --cluster-name=kubernetes --root-ca-file=/etc/kubernetes/pki/ca.pem --service-account-private-key-file=/etc/kubernetes/pki/apiserver-key.pem --cluster-signing-cert-file=/etc/kubernetes/pki/ca.pem --cluster-signing-key-file=/etc/kubernetes/pki/ca-key.pem --insecure-experimental-approve-all-kubelet-csrs-for-group=system:kubelet-bootstrap
[root@r03s09 ~]# ps aux|grep kube-dns
root      14319  0.0  0.0 112648   968 pts/0    S+   13:32   0:00 grep --color=auto kube-dns
root     167414  0.1  0.0  57944 28188 ?        Ssl  Jul24  63:31 /kube-dns --domain=cluster.local --dns-port=10053
[root@r03s09 ~]# hostname
r03s09
[root@r03s09 ~]# hostname -f
r03s09
[root@r03s09 ~]# kubectl get nodes
NAME      STATUS    AGE
r03s09    Ready     36d
r03s10    Ready     35d
r03s11    Ready     36d
```
During my test, the 5 seconds overhead caused by querying host name from IP address happens not only in the scheduler back-end, but also in fetching HDFS file.

Do you met such issues in your environment?

tangzhankun commented:
@kimoonkim 
A finding might be helpful to analyze the issue, if I "docker run" with `"--add-host"` option, the "test_inet.java" runs fast.
```java
import java.net.InetAddress;
public class test_inet {
  public static void main(String[] args) {
    try{
      System.out.println(InetAddress.getByName("172.16.3.70").getCanonicalHostName());
    }catch(Exception e) {
      e.printStackTrace();
    }
  }
}
```
```bash
[root@r03s09 ~]# docker run --rm -it --add-host="r03s10:172.16.3.70" spark-base:debug /bin/bash
bash-4.3# 
bash-4.3# 
bash-4.3# vi test_inet.java
bash-4.3# javac test_inet.java 
bash-4.3# time java test_inet
r03s10

real	0m0.127s
user	0m0.107s
sys	0m0.045s
```

But it seems I cannot add this option in k8s v1.4.0(with HostAlises of v1.7?) to do double-check, any idea?

kimoonkim commented:
> During my test, the 5 seconds overhead caused by querying host name from IP address happens not only in the scheduler back-end, but also in fetching HDFS file.

The five sec delay might still happen even if your setup is completely correct, in case the host to task mapping is empty. That's one of the reasons why the bug is quite bad. The code path cannot determine when to stop and ends up issuing the DNS request.

I do not know much about `--add-host` option.

Your hostname setup is all short name based? How about your bare-metal DNS setup. Doesn't it have a domain name? I suspect you want to change the hostnames to match what the bare-metal DNS resolves to.

> [root@r03s09 ~]# hostname
r03s09
[root@r03s09 ~]# hostname -f
r03s09
[root@r03s09 ~]# kubectl get nodes
NAME      STATUS    AGE
r03s09    Ready     36d
r03s10    Ready     35d
r03s11    Ready     36d

tangzhankun commented:
@kimoonkim Yeah. I'm using short name at present. 
As you said, I changed the DNS servers in the physical hosts' /etc/resolv.conf. The overhead is reduced to less than 1s. Thanks.

kimoonkim commented:
@tangzhankun Glad you resolved the problem. This hostname/DNS setup is tricky :-)
